### ISSUE TITLE
importWorkboxFrom: 'local' bypasses the webpack asset pipeline

### ISSUE THREAD

**Issue by jeffposnick** [GoogleChrome/workbox#1338]
**Library Affected**:
*workbox-webpack-plugin*

Moving over @raejin's [comment](https://github.com/GoogleChrome/workbox/pull/1290#issuecomment-369423980) into a standalone issue:

> After testing out v3.0.0-beta.1, I think this is not compatible with static asset that gets served from a different domain. By tracing the source a bit, we are copying the file on the disk instead of copying the file to the Webpack compilation.
> 
> https://github.com/raejin/workbox/blob/17c38de990585d7076b03ecca55d33b3166e53e9/packages/workbox-build/src/lib/copy-workbox-libraries.js#L78-L79
> 
> What do you guys think of a different implementation on copying on disk versus Webpack compilation?

**Comment by jeffposnick** [comment]
You're correct that we end up copying files using the "real" filesystem rather than adding the files to the webpack compilation. There's a discussion of this in the codebase:

https://github.com/GoogleChrome/workbox/blob/17c38de990585d7076b03ecca55d33b3166e53e9/packages/workbox-webpack-plugin/src/lib/get-workbox-sw-imports.js#L36-L50

The alternative is not being able to share that `copyWorkboxLibraries()` logic from `workbox-build` and implementing something equivalent that just works for webpack. That's possible, but I want to understand exactly *how* this is breaking compatibility for your setup.

The code currently takes `compilation.options.output.publicPath` into account when coming up with the URL used for the import,  which I was under the impression was the "standard" way of specifying "custom" serving URLs. Can you explain more about what constraints you have?

Also, given that we're close to the v3.0.0 release, it would be good to understand how much of a "blocker" you consider this issue.

**Comment by raejin** [comment]
@jeffposnick

I should have clarified that this behavior will make it inconsistent in development, specifically for our Webpack setup which relies on a separate Webpack dev server (which will serve the asset from a separate subdomain).

Here is a simple illustration of what the structure looks like:

```
├── app // all the source code for JS
└── public
    ├── service-worker.js // on disk, served from a non Webpack dev server domain
    └── webpackFolder
        ├── precache-manifest.js  // In development, this doesn't live on disk, which will served from a separate domain
        ├── workbox-generated-service-worker-script.js // In development, this doesn't live on disk, which will served from a separate domain
        └── workbox-v3.0.0-beta.1 // on disk, served from a non Webpack dev server domain
```

This is not a blocker at all provided that the solution definitely works in production. Prior to beta.1, I already had a solution that copies on behave of Webpack asset pipeline for development && production consistency.

Thank you for looking into this!

**Comment by jeffposnick** [comment]
Thanks for the additional information, @raejin.

Given that, I'm not going to consider this within scope for the upcoming v3 release, but will leave this issue open to track future refactoring to use the webpack asset pipeline instead of the local filesystem when in `importWorkboxFrom: 'local'` mode.

**Comment by frlinw** [comment]
In 3.1.0 and dev only, precache-manifest is copied in `importsDirectory` but workbox-v3.1.0 is not
```javascript
    new InjectManifest({
      swSrc: `${rootPath}/src/sw.js`,
      importWorkboxFrom: "local",
      importsDirectory: "static/lib/workbox",
      include: [/\.html$/, /\.js$/, /\.css$/, /\.woff2$/]
    })
```
webpack 4.5.0
webpack-serve 0.3.0

**Comment by jeffposnick** [comment]
In Workbox v5, any assets created by `GenerateSW` are added directly to `compilation.assets` and rather than written to the file system.

(We also generate a single, custom runtime bundle of the Workbox code you actually use, rather than copy over the full set of Workbox libraries to the output directory.)

**Comment by jeffposnick** [comment]
This should be addressed by the current Workbox v5.0.0 alpha.
